lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java

```diff
@@ -16,30 +16,138 @@
+     *
+     * @throws IllegalArgumentException if location is not valid for this file system
+     * @throws IOException if the file does not exist or was not deleted.
+     */
     void deleteFile(String location)
             throws IOException;
 
     /**
-     * Delete files in batches, possibly non-atomically.
-     * If an error occurs, some files may have been deleted.
+     * Delete specified files. This operation is <b>not</b> required to be atomic, so if an error
+     * occurs, all, some, or, none files may be deleted. This operation may be faster than simply
+     * looping over the location as some file systems support batch delete operations natively.
+     *
+     * @throws IllegalArgumentException if location is not valid for this file system
+     * @throws IOException if a file does not exist or was not deleted.
      */
-    void deleteFiles(Collection<String> locations)
-            throws IOException;
+    default void deleteFiles(Collection<String> locations)
+            throws IOException
+    {
+        for (String location : locations) {
+            deleteFile(location);
+        }
+    }
 
+    /**
+     * Deletes all files and directories within the specified directory recursively, and deletes
+     * the directory itself. If the location does not exist, this method is a noop. If the location
+     * does not have a path, all files and directories in the file system are deleted.
+     * <p>
+     * For hierarchical file systems (e.g. HDFS), if the path is not a directory, an exception is
+     * raised.
+     * <p>
+     * For blob file systems (e.g., S3), if the location does not end with a slash one is appended,
+     * and all blobs that start with that prefix are deleted.
+     * <p>
+     * If this operation fails, some, none, or all of the contents of the directory may
+     * have been deleted.
+     *
+     * @param location the directory to delete
+     * @throws IllegalArgumentException if location is not valid for this file system
+     */
     void deleteDirectory(String location)
             throws IOException;
 
+    /**
+     * Rename source to target without overwriting target.  This method is not required
+     * to be atomic, but it is required that if an error occurs the source, target, or both
+     * must exist with the data from the source.  The last modified time, may or may not be
+     * preserved.
+     *
+     * @throws IllegalArgumentException if either location is not valid for this file system
+     */
     void renameFile(String source, String target)
             throws IOException;
 
+    /**
+     * Lists all files within the specified directory recursively. The location can be empty,
+     * listing all files in the file system, otherwise the location must end with a slash. If the
+     * location does not exist, and empty iterator is returned.
+     * <p>
+     * For hierarchical file systems (e.g. HDFS), if the path is not a directory, an exception is
```

Comment:
We could remove the "e.g. HDFS" since we discuss hierarchical file systems in the introduction